src/shammath/include/shammath/crystalLattice_smap_sphere.hpp

```diff
@@ -0,0 +1,502 @@
+// -------------------------------------------------------//
+//
+// SHAMROCK code for hydrodynamics
+// Copyright (c) 2021-2025 Timothée David--Cléris <tim.shamrock@proton.me>
+// SPDX-License-Identifier: CeCILL Free Software License Agreement v2.1
+// Shamrock is licensed under the CeCILL 2.1 License, see LICENSE for more information
+//
+// -------------------------------------------------------//
+
+#pragma once
+
+/**
+ * @file crystalLattice_smap_sphere.hpp
+ * @author Timothée David--Cléris (tim.shamrock@proton.me)
+ * @author David Fang (david.fang@ikmail.com)
+ * @brief
+ *
+ */
+
+#include "shambase/aliases_int.hpp"
+#include "shambase/constants.hpp"
+#include "shambackends/sycl_utils.hpp"
+#include "shambackends/vec.hpp"
+#include "shamcomm/logs.hpp"
+#include "shammath/CoordRange.hpp"
+#include "shammath/DiscontinuousIterator.hpp"
+#include "shammath/LatticeError.hpp"
+#include "shammath/integrator.hpp"
+#include <hipSYCL/sycl/libkernel/builtins.hpp>
+#include <array>
+#include <functional>
+#include <utility>
+#include <vector>
+
+namespace shammath {
+
+    /**
+     * @brief utility for generating stretched spherical HCP crystal lattices given a density
+     * profile
+     *
+     * @tparam Tvec position vector type
+     */
+    template<class Tvec>
+    class LatticeHCP_smap_sphere {
+
+        public:
+        static constexpr u32 dim = 3;
+
+        static_assert(
+            dim == shambase::VectorProperties<Tvec>::dimension, "this lattice exist only in dim 3");
+
+        using Tscal               = shambase::VecComponent<Tvec>;
+        static constexpr Tscal pi = shambase::constants::pi<Tscal>;
+
+        static Tscal stretchindiv(
+            Tscal r,
+            std::function<Tscal(Tscal)> rhoprofile,
+            Tscal integral_profile,
+            Tscal rmin,
+            Tscal rmax,
+            Tvec center) {
+            constexpr Tscal tol     = 1e-9;
+            constexpr u32 maxits    = 200;
+            constexpr u32 maxits_nr = 20;
+            constexpr u32 ngrid     = 2000;
+            Tscal step              = (rmax - rmin) / ngrid;
+
+            u32 its       = 0;
+            Tscal initpos = r;
+            Tscal newr    = r;
+            Tscal prevr   = r;
+
+            std::function<Tscal(Tscal)> rhodS = [&rhoprofile](Tscal r) -> Tscal {
```

Comment:
```suggestion
            auto rhodS = [&rhoprofile](Tscal r) -> Tscal {
```
The lambdas can be optimized away, but forcing the conversion to a std::function (which is not equivalent) may degrade the performance. It is better to keep lambda as `auto` in general